tsfc/loopy.py

```diff
@@ -15,11 +15,41 @@
 import pymbolic.primitives as p
 from loopy.symbolic import SubArrayRef
 
-from pytools import UniqueNameGenerator
+from pytools import UniqueNameGenerator, ImmutableRecord
 
 from tsfc.parameters import is_complex
 
 from contextlib import contextmanager
+from petsc4py import PETSc
+
+
+def profile_insns(kernel_name, instructions):
+    if PETSc.Log.getActive():
+        # Petsc functions
+        start_event = "PetscLogEventBegin"
+        end_event = "PetscLogEventEnd"
+        register_event = "PetscLogEventRegister"
+        # The variables
+        event_id_var_name = "event_id_" + kernel_name
+        event_name = "Log_Event_" + kernel_name
+        # Logging registration # TODO offport this so we don't register every single time
+        prepend = [lp.CInstruction("", "PetscLogEvent "+event_id_var_name+";"),
+                   lp.CInstruction("", register_event+"(\""+event_name+"\", PETSC_OBJECT_CLASSID, &"+event_id_var_name+");")]
```

Comment:
Depending on if we decide that the faffing with declaring the events outside the C kernels in python and setting the variables in the DLLs is worth it, we want to remove declaration in 36 and the registration in 37 and instead declare the event in the preamble.